File: src/main/java/com/akagiyui/drive/service/impl/UserServiceImpl.java
package com.akagiyui.drive.service.impl;

import cn.hutool.core.util.RandomUtil;
import com.akagiyui.common.ResponseEnum;
import com.akagiyui.common.exception.CustomException;
import com.akagiyui.drive.component.CacheConstants;
import com.akagiyui.drive.component.RedisCache;
import com.akagiyui.drive.entity.Role;
import com.akagiyui.drive.entity.User;
import com.akagiyui.drive.model.LoginUserDetails;
import com.akagiyui.drive.model.filter.UserFilter;
import com.akagiyui.drive.model.request.AddUserRequest;
import com.akagiyui.drive.model.request.EmailVerifyCodeRequest;
import com.akagiyui.drive.model.request.RegisterConfirmRequest;
import com.akagiyui.drive.model.request.UpdateUserInfoRequest;
import com.akagiyui.drive.repository.UserRepository;
import com.akagiyui.drive.service.ConfigService;
import com.akagiyui.drive.service.MailService;
import com.akagiyui.drive.service.RoleService;
import com.akagiyui.drive.service.UserService;
import jakarta.annotation.Resource;
import jakarta.persistence.criteria.Predicate;
import kotlin.NotImplementedError;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.context.annotation.Lazy;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;


/**
 * 用户服务实现类
 *
 * @author AkagiYui
 */
@Service
@Slf4j
public class UserServiceImpl implements UserService {

    @Resource
    UserRepository repository;

    @Resource
    RedisCache redisCache;

    @Value("${application.email.verify.timeout}")
    private long emailVerifyTimeout;

    @Resource
    MailService mailService;

    @Resource
    @Lazy
    PasswordEncoder passwordEncoder;

    @Resource
    private ConfigService configService;

    @Resource
    private RoleService roleService;

    @Override
    @Cacheable(value = CacheConstants.USER_BY_ID, key = "#id")
    public @NotNull User findUserById(@NotNull String id) {
        return repository.findById(id).orElseThrow(() -> new CustomException(ResponseEnum.NOT_FOUND));
    }

    @Override
    public @NotNull List<User> findUserByIds(@NotNull List<String> ids) {
        return repository.findAllById(ids);
    }

    private User findUserByIdWithCache(String id) {
        return repository.findById(id).orElseThrow(() -> new CustomException(ResponseEnum.NOT_FOUND));
    }

    @Override
    public @NotNull User register(@NotNull User user) {
        throw new NotImplementedError("register");
    }

    @Override
    @Cacheable(cacheNames = CacheConstants.USER_PAGE, key = "{#index, #size, #filter}")
    public @NotNull Page<User> find(int index, int size, UserFilter filter) {
        Pageable pageable = PageRequest.of(index, size);

        // 条件查询
        Specification<User> specification = (root, query, cb) -> {
            if (filter != null && StringUtils.hasText(filter.getExpression())) {
                String queryString = "%" + filter.getExpression() + "%";
                Predicate usernamePredicate = cb.like(root.get("username"), queryString);
                Predicate nicknamePredicate = cb.like(root.get("nickname"), queryString);
                Predicate emailPredicate = cb.like(root.get("email"), queryString);
                return cb.or(usernamePredicate, nicknamePredicate, emailPredicate);
            }
            return null;
        };

        return repository.findAll(specification, pageable);
    }

    @Override
    @Cacheable(cacheNames = CacheConstants.USER_LIST)
    public @NotNull List<User> find() {
        return repository.findAll();
    }

    @Override
    @CacheEvict(cacheNames = {
        CacheConstants.USER_BY_ID,
        CacheConstants.USER_PAGE,
        CacheConstants.USER_LIST,
        CacheConstants.USER_EXIST,
    }, allEntries = true)
    public @NotNull User addUser(AddUserRequest user) {
        if (repository.existsByUsername(user.getUsername())) {
            throw new CustomException(ResponseEnum.USER_EXIST);
        }

        User realUser = user.toUser();
        if (StringUtils.hasText(user.getEmail())) {
            if (repository.existsByEmail(user.getEmail())) {
                throw new CustomException(ResponseEnum.EMAIL_EXIST);
            }
            realUser.setEmail(user.getEmail());
        }
        if (StringUtils.hasText(user.getNickname())) {
            realUser.setNickname(user.getNickname());
        }

        realUser.setPassword(encryptPassword(user.getUsername(), user.getPassword()));
        realUser.setDisabled(false);
        realUser.setRoles(roleService.getAllDefaultRoles());

        repository.save(realUser);
        return realUser;
    }

    @Override
    @CacheEvict(cacheNames = {
        CacheConstants.USER_BY_ID,
        CacheConstants.USER_DETAILS,
        CacheConstants.USER_LOGIN_DETAILS,
        CacheConstants.USER_PAGE,
        CacheConstants.USER_LIST,
        CacheConstants.USER_EXIST,
    }, allEntries = true)
    public void delete(@NotNull String id) {
        if (!repository.existsById(id)) {
            throw new CustomException(ResponseEnum.NOT_FOUND);
        }
        repository.deleteById(id);
    }

    @Override
    @Cacheable(cacheNames = CacheConstants.USER_EXIST, key = "#username")
    public boolean isExist(@NotNull String username) {
        return repository.existsById(username);
    }

    @Override
    public @NotNull User getUser() {
        // 从 SecurityContextHolder 中获取用户信息
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        LoginUserDetails userDetails = (LoginUserDetails) authentication.getPrincipal();
        return userDetails.getUser();
    }

    @Override
    @Cacheable(cacheNames = CacheConstants.USER_LOGIN_DETAILS, key = "#userId")
    @Transactional
    public @NotNull LoginUserDetails getUserDetails(@NotNull String userId) {
        User user = findUserById(userId);
        return new LoginUserDetails(user);
    }

    @Override
    public void sendEmailVerifyCode(@NotNull EmailVerifyCodeRequest verifyRequest) {
        if (!configService.isRegisterEnabled()) {
            throw new CustomException(ResponseEnum.REGISTER_DISABLED);
        }

        // 检查该邮箱是否在 redis 中等待验证
        String redisKey = "emailVerifyCode:" + verifyRequest.getEmail();
        if (Objects.equals(Boolean.TRUE, redisCache.hasKey(redisKey))) {
            throw new CustomException(ResponseEnum.EMAIL_EXIST);
        }
        // 检查该邮箱是否已经注册
        if (repository.existsByEmail(verifyRequest.getEmail())) {
            throw new CustomException(ResponseEnum.EMAIL_EXIST);
        }
        // 检查用户名是否已经注册
        if (repository.existsByUsername(verifyRequest.getUsername())) {
            throw new CustomException(ResponseEnum.EMAIL_EXIST);
        }
        // 生成验证码
        String verifyCode = RandomUtil.randomNumbers(6);
        redisCache.set(redisKey, verifyCode);
        redisCache.expire(redisKey, emailVerifyTimeout, TimeUnit.MINUTES);
        mailService.sendEmailVerifyCode(verifyRequest.getEmail(), verifyCode, emailVerifyTimeout);
        // 将注册信息存入 redis
        String registerInfoKey = "registerInfo:" + verifyRequest.getEmail();
        redisCache.set(registerInfoKey, verifyRequest);
        redisCache.expire(registerInfoKey, emailVerifyTimeout + 1, TimeUnit.MINUTES);
    }

    @Override
    public void confirmRegister(RegisterConfirmRequest registerConfirmRequest) {
        // 从 redis 取回验证码
        String redisKey = "emailVerifyCode:" + registerConfirmRequest.getEmail();
        String verifyCode = redisCache.get(redisKey);
        if (verifyCode == null) {
            throw new CustomException(ResponseEnum.VERIFY_CODE_NOT_FOUND);
        }
        // 检查验证码是否正确
        if (!Objects.equals(registerConfirmRequest.getVerifyCode(), verifyCode)) {
            throw new CustomException(ResponseEnum.VERIFY_CODE_NOT_FOUND);
        }
        // 从 redis 取回用户注册信息
        EmailVerifyCodeRequest verifyRequest = redisCache.get("registerInfo:" + registerConfirmRequest.getEmail());
        if (verifyRequest == null) {
            log.error("Register info not found: {}", registerConfirmRequest.getEmail());
            throw new CustomException(ResponseEnum.VERIFY_CODE_NOT_FOUND);
        }
        // 转换为用户对象
        User user = new User();
        user.setUsername(verifyRequest.getUsername());
        user.setPassword(encryptPassword(verifyRequest.getUsername(), verifyRequest.getPassword()));
        user.setEmail(verifyRequest.getEmail());
        user.setDisabled(false);
        try {
            repository.save(user);
        } finally {
            // 删除 redis 中的验证码和注册信息
            redisCache.delete(redisKey);
            redisCache.delete("registerInfo:" + registerConfirmRequest.getEmail());
        }
    }

    @Override
    public @NotNull String encryptPassword(@NotNull String username, @NotNull String password) {
        return encryptPassword(username, password, false);
    }

    @Override
    public @NotNull String encryptPassword(@NotNull String username, @NotNull String password, boolean raw) {
        @SuppressWarnings("UnnecessaryLocalVariable")
        // 密码加密核心
        String encode = password;
        if (raw) {
            return encode;
        }
        return passwordEncoder.encode(encode);
    }

    @Override
    @CacheEvict(cacheNames = {
        CacheConstants.USER_BY_ID,
        CacheConstants.USER_DETAILS,
        CacheConstants.USER_LOGIN_DETAILS,
        CacheConstants.USER_PAGE,
        CacheConstants.USER_LIST,
        CacheConstants.USER_EXIST,
    }, allEntries = true)
    public void updateInfo(UpdateUserInfoRequest userInfo) {
        User user = getUser();
        if (StringUtils.hasText(userInfo.getNickname())) {
            user.setNickname(userInfo.getNickname());
        }
        if (StringUtils.hasText(userInfo.getEmail())) {
            user.setEmail(userInfo.getEmail());
        }
        repository.save(user);
    }

    @Override
    @Transactional
    public @NotNull Set<String> getPermission() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        LoginUserDetails userDetails = (LoginUserDetails) authentication.getPrincipal();
        return userDetails.getPermissions();
    }

    @Override
    public @NotNull Set<String> getRole() {
        User user = getUser();
        return user.getRoles().stream().map(Role::getId).collect(Collectors.toSet());
    }

    @Override
    @CacheEvict(cacheNames = {
        CacheConstants.USER_BY_ID,
        CacheConstants.USER_DETAILS,
        CacheConstants.USER_LOGIN_DETAILS,
        CacheConstants.USER_PAGE,
        CacheConstants.USER_LIST,
        CacheConstants.USER_EXIST,
    }, allEntries = true)
    public void disable(@NotNull String id, boolean disabled) {
        User user = findUserByIdWithCache(id);
        // todo 检查是否为超级管理员
        user.setDisabled(disabled);
        repository.save(user);
    }

    @Override
    public void resetPassword(@NotNull String id, @NotNull String newPassword) {
        User user = findUserByIdWithCache(id);
        user.setPassword(encryptPassword(user.getUsername(), newPassword));
        repository.save(user);
    }

    @Override
    @Transactional
    public void addRoles(@NotNull String userId, @NotNull Set<String> id) {
        User user = findUserByIdWithCache(userId);
        Set<Role> roles = roleService.find(id);
        user.getRoles().addAll(roles);
        repository.save(user);
    }

    @Override
    @Transactional
    public void removeRoles(@NotNull String userId, @NotNull Set<String> id) {
        User user = findUserByIdWithCache(userId);
        Set<Role> roles = roleService.find(id);
        user.getRoles().removeAll(roles);
        repository.save(user);
    }

    @Override
    public void updateInfo(@NotNull String id, UpdateUserInfoRequest userInfo) {
        User user = findUserByIdWithCache(id);
        if (StringUtils.hasText(userInfo.getNickname())) {
            user.setNickname(userInfo.getNickname());
        }
        if (StringUtils.hasText(userInfo.getEmail())) {
            user.setEmail(userInfo.getEmail());
        }
        if (StringUtils.hasText(userInfo.getPassword())) {
            user.setPassword(encryptPassword(user.getUsername(), userInfo.getPassword()));
        }
        repository.save(user);
    }

    @Override
    @Transactional
    public @NotNull Set<Role> getRoles(@NotNull String id) {
        User user = findUserByIdWithCache(id);
        return new HashSet<>(user.getRoles());
    }

    /**
     * 根据用户名获取用户信息
     *
     * @param loginUsernameParam 登录username参数
     */
    @Override
    @Cacheable(cacheNames = CacheConstants.USER_DETAILS, key = "#loginUsernameParam")
    @Transactional
    public UserDetails loadUserByUsername(String loginUsernameParam) throws UsernameNotFoundException {
        User user = repository.getFirstByUsernameOrEmail(loginUsernameParam);
        if (user == null) {
            throw new UsernameNotFoundException("Username or password error");
        }
        return new LoginUserDetails(user);
    }
}

File: app/src/main/kotlin/com/akagiyui/drive/service/impl/UserServiceImpl.kt
package com.akagiyui.drive.service.impl

import cn.hutool.core.util.RandomUtil
import com.akagiyui.common.ResponseEnum
import com.akagiyui.common.delegate.LoggerDelegate
import com.akagiyui.common.exception.CustomException
import com.akagiyui.common.token.TokenTemplate
import com.akagiyui.common.utils.hasText
import com.akagiyui.drive.component.RedisCache
import com.akagiyui.drive.entity.Role
import com.akagiyui.drive.entity.User
import com.akagiyui.drive.model.AddUserModel
import com.akagiyui.drive.model.CacheConstants
import com.akagiyui.drive.model.UserFilter
import com.akagiyui.drive.model.cache.EmailRegisterInfo
import com.akagiyui.drive.model.request.user.UpdateUserInfoRequest
import com.akagiyui.drive.repository.UserRepository
import com.akagiyui.drive.service.*
import jakarta.annotation.Resource
import jakarta.persistence.criteria.Predicate
import org.jetbrains.annotations.NotNull
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.Cacheable
import org.springframework.context.annotation.Lazy
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.domain.Specification
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.concurrent.TimeUnit
import java.util.stream.Collectors

/**
 * 用户服务实现类
 *
 * @author AkagiYui
 */
@Service
class UserServiceImpl(
    private val repository: UserRepository,
    private val redisCache: RedisCache,
    private val mailService: MailService,
    private val smsService: SmsService,
    private val settingService: SettingService,
    private val roleService: RoleService,
    private val ipRegionService: IpRegionService,
    private val tokenTemplate: TokenTemplate,
    private val passwordEncoder: PasswordEncoder,
) : UserService {
    private val log by LoggerDelegate()
    private var emailVerifyTimeout: Long = settingService.mailVerifyCodeTimeout.toLong()

    @Cacheable(value = [CacheConstants.USER_BY_ID], key = "#id")
    override fun getUserById(id: String): User {
        return repository.findById(id).orElseThrow { CustomException(ResponseEnum.NOT_FOUND) }
    }

    override fun getUserByPhone(phone: String): User? {
        return repository.findByPhone(phone)
    }

    override fun getUserByUsernameOrEmailOrPhone(query: String): User? {
        return repository.getFirstByUsernameOrEmailOrPhone(query)
    }

    override fun getUsersByIds(@NotNull ids: List<String>): List<User> {
        return repository.findAllById(ids)
    }

    private fun findUserByIdWithCache(id: String): User {
        return repository.findById(id).orElseThrow { CustomException(ResponseEnum.NOT_FOUND) }
    }

    @Cacheable(cacheNames = [CacheConstants.USER_PAGE], key = "{#index, #size, #userFilter }")
    override fun getUsers(index: Int, size: Int, userFilter: UserFilter?): Page<User> {
        val pageable: Pageable = PageRequest.of(index, size)

        val specification: Specification<User> = Specification { root, _, cb ->
            if (userFilter != null && userFilter.expression.hasText()) {
                val queryString = "%${userFilter.expression}%"
                val usernamePredicate: Predicate = cb.like(root["username"], queryString)
                val nicknamePredicate: Predicate = cb.like(root["nickname"], queryString)
                val emailPredicate: Predicate = cb.like(root["email"], queryString)
                cb.or(usernamePredicate, nicknamePredicate, emailPredicate)
            } else null
        }

        return repository.findAll(specification, pageable)
    }

    @Cacheable(cacheNames = [CacheConstants.USER_LIST])
    override fun getUsers(): List<User> {
        return repository.findAll()
    }

    @CacheEvict(
        cacheNames = [
            CacheConstants.USER_BY_ID,
            CacheConstants.USER_PAGE,
            CacheConstants.USER_LIST,
            CacheConstants.USER_EXIST,
        ],
        allEntries = true,
    )
    override fun addUser(request: AddUserModel): User {
        val entity = User()
        entity.roles = roleService.getAllDefaultRoles() // 添加默认角色

        val currentTimestamp = System.currentTimeMillis()
        request.email.hasText {
            if (repository.existsByEmail(it)) {
                throw CustomException(ResponseEnum.EMAIL_EXIST)
            }
            entity.email = it
            entity.nickname = it
            entity.username = "email:$it:$currentTimestamp"
        }
        request.phone.hasText {
            if (repository.existsByPhone(it)) {
                throw CustomException(ResponseEnum.PHONE_EXIST)
            }
            entity.phone = it
            entity.nickname = it
            entity.username = "phone:$it:$currentTimestamp"
        }
        // 设置用户名，优先级最高，覆盖前面的设置
        request.username.hasText {
            if (repository.existsByUsername(it)) {
                throw CustomException(ResponseEnum.USER_EXIST)
            }
            entity.username = it
        }

        // 密码加密
        request.password.hasText { entity.password = encryptPassword(entity.username, it) }
        // 设置昵称
        request.nickname.hasText { entity.nickname = it }
        return repository.save(entity)
    }

    @CacheEvict(
        cacheNames = [
            CacheConstants.USER_BY_ID,
            CacheConstants.USER_DETAILS,
            CacheConstants.USER_PAGE,
            CacheConstants.USER_LIST,
            CacheConstants.USER_EXIST,
        ],
        allEntries = true,
    )
    override fun deleteUser(id: String) {
        if (!repository.existsById(id)) {
            throw CustomException(ResponseEnum.NOT_FOUND)
        }
        repository.deleteById(id)
        // todo 删除用户文件关联
    }

    @Cacheable(cacheNames = [CacheConstants.USER_EXIST], key = "#id")
    override fun isUserExist(id: String): Boolean {
        return repository.existsById(id)
    }


    override fun registerByEmail(email: String, password: String) {
        if (!settingService.registerEnabled) {
            throw CustomException(ResponseEnum.REGISTER_DISABLED)
        }

        // 检查该邮箱是否在 redis 中等待验证
        val cacheKey = buildEmailRegisterCacheKey(email)
        if (cacheKey in redisCache) {
            throw CustomException(ResponseEnum.EMAIL_EXIST)
        }
        // 检查该邮箱是否已经注册
        if (repository.existsByEmail(email)) {
            throw CustomException(ResponseEnum.EMAIL_EXIST)
        }
        // 生成验证码
        val otp = RandomUtil.randomNumbers(6)
        mailService.sendEmailVerifyCode(email, otp, emailVerifyTimeout)
        // 将注册信息存入 redis
        val cacheModel = EmailRegisterInfo(email, password, otp)
        redisCache[cacheKey, emailVerifyTimeout + 1, TimeUnit.MINUTES] = cacheModel
    }

    private fun buildEmailRegisterCacheKey(email: String): String {
        return "emailRegisterInfo:$email"
    }

    /**
     * 用户服务，自我注入，以触发 AOP 代理
     */
    @Resource
    @Lazy
    lateinit var selfProxy: UserService

    override fun confirmRegister(email: String, otp: String) {
        // 从 redis 取回用户注册信息
        val cacheKey = buildEmailRegisterCacheKey(email)
        val cacheModel = redisCache.get<EmailRegisterInfo>(cacheKey)
            ?: throw CustomException(ResponseEnum.VERIFY_CODE_NOT_FOUND)
        // 检查验证码是否正确
        if (otp != cacheModel.otp) {
            throw CustomException(ResponseEnum.VERIFY_CODE_NOT_FOUND)
        }
        // 添加用户
        selfProxy.addUser(AddUserModel().apply {
            password = cacheModel.password
            this.email = cacheModel.email
        })
        // 删除 redis 中的注册信息
        redisCache.delete(cacheKey)
    }

    /**
     * 加密密码
     * @param username 用户名
     * @param password 密码明文
     * @return 密码密文
     */
    private fun encryptPassword(username: String, password: String): String {
        return encryptPassword(username, password, false)
    }

    /**
     * 加密密码
     * @param username 用户名
     * @param password 密码明文
     * @param raw 是否不通过加密器加密
     * @return 密码密文
     */
    private fun encryptPassword(username: String, password: String, raw: Boolean): String {
        val encode = password // 密码加密核心
        if (raw) {
            return encode
        }
        return passwordEncoder.encode(encode)
    }


    @Transactional
    override fun getPermission(user: User): Set<String> {
        return user.roles
            .asSequence()
            .map { it.permissions }
            .flatten()
            .map { it.name }
            .toSet()
    }

    override fun getRole(user: User): Set<String> {
        return user.roles.stream().map(Role::id).collect(Collectors.toSet())
    }

    @CacheEvict(
        cacheNames = [
            CacheConstants.USER_BY_ID,
            CacheConstants.USER_DETAILS,
            CacheConstants.USER_PAGE,
            CacheConstants.USER_LIST,
            CacheConstants.USER_EXIST,
        ],
        allEntries = true,
    )
    override fun disable(id: String, disabled: Boolean) {
        val user = findUserByIdWithCache(id)
        // todo 检查是否为超级管理员
        user.disabled = disabled
        repository.save(user)
    }

    override fun resetPassword(id: String, newPassword: String) {
        val user = findUserByIdWithCache(id)
        user.password = encryptPassword(user.username, newPassword)
        repository.save(user)
    }

    @Transactional
    override fun addRoles(userId: String, id: Set<String>) {
        val user = findUserByIdWithCache(userId)
        val roles = roleService.find(id)
        user.roles.addAll(roles)
        repository.save(user)
    }

    @Transactional
    override fun removeRoles(userId: String, id: Set<String>) {
        val user = findUserByIdWithCache(userId)
        val roles = roleService.find(id)
        user.roles.removeAll(roles)
        repository.save(user)
    }

    @CacheEvict(
        cacheNames = [
            CacheConstants.USER_BY_ID,
            CacheConstants.USER_DETAILS,
            CacheConstants.USER_PAGE,
            CacheConstants.USER_LIST,
        ],
        allEntries = true,
    )
    override fun updateInfo(id: String, userInfo: UpdateUserInfoRequest) {
        val user = findUserByIdWithCache(id)
        userInfo.nickname.hasText { user.nickname = it }
        userInfo.email.hasText { user.email = it }
        userInfo.phone.hasText { user.phone = it }
        userInfo.password.hasText { user.password = encryptPassword(user.username, it) }
        repository.save(user)
    }

    @Transactional
    override fun getRoles(id: String): Set<Role> {
        val user = findUserByIdWithCache(id)
        return HashSet(user.roles)
    }


}
